src/cookies.ts

```diff
@@ -0,0 +1,129 @@
+const memoizedCookies: Map<string, string> = new Map<string, string>();
+
+const getCookieValues = (name: string) => {
+	const nameEq = `${name}=`;
+	const cookies = document.cookie.split(';');
+
+	return cookies.reduce((acc: string[], cookie: string) => {
+		const cookieTrimmed: string = cookie.trim();
+		if (cookieTrimmed.startsWith(nameEq)) {
+			acc.push(
+				cookieTrimmed.substring(nameEq.length, cookieTrimmed.length),
+			);
+		}
+
+		return acc;
+	}, []);
+};
+
+// subset of https://github.com/guzzle/guzzle/pull/1131
+const isValidCookieValue = (name: string) =>
+	!/[()<>@,;"\\/[\]?={} \t]/g.test(name);
+
+const getShortDomain = ({ isCrossSubdomain = false } = {}) => {
+	const domain = document.domain || '';
+
+	if (domain === 'localhost' || window.guardian?.config?.page?.isPreview) {
+		return domain;
+	}
+
+	// Trim any possible subdomain (will be shared with supporter, identity, etc)
+	if (isCrossSubdomain) {
+		return ['', ...domain.split('.').slice(-2)].join('.');
+	}
+	// Trim subdomains for prod (www.theguardian), code (m.code.dev-theguardian) and dev (dev.theguardian, m.thegulocal)
+	return domain.replace(/^(www|m\.code|dev|m)\./, '.');
+};
+
+const getDomainAttribute = ({ isCrossSubdomain = false } = {}) => {
+	const shortDomain = getShortDomain({ isCrossSubdomain });
+	return shortDomain === 'localhost' ? '' : ` domain=${shortDomain};`;
+};
+
+/**
+ * Set a cookie. If it's been memoized it won't retrieve it again.
+ * @param {string} name - the cookie’s name.
+ * @param {string} value - the cookie’s value.
+ * @param {boolean} shouldMemoize - set to true if you want to memoize it, default false.
+ * @param {number} daysToLive - expiry date will be calculated mased on the daysToLive
+ * @param {boolean} isCrossSubdomain - specify if it's a cross subdomain cookie, default false
+ */
+export const setCookie = (
+	name: string,
+	value: string,
+	daysToLive?: number,
+	isCrossSubdomain = false,
+	shouldMemoize = false,
+): void => {
```

Comment:
have you considered using a params object here? it would enable to you write (and later read) uses like this:

```js
setCookie({ name: "a", value: "b", isCrossSubdomain: true })
setCookie({ name: "a", value: "b", daysToLive: 1, shouldMemoize: true })
```

here's a [example in typescript](https://www.typescriptlang.org/play?#code/MYewdgzgLgBBCmUDCIQGsCW8YF4YAoBvAKBhjAEMBbeAGlJgDcKAbAVzoYBMKBPCACogAMhkacyGCEgBOICBADKbAEZcQVChjC4YAM1YJ6ZCAAsQbFlwCy8KiAwAvbHgMsjxAL4AuGCTKUNL7QMtoA5gDcDMzs8MFQoWCR3HyCImLwAPy+YGxUKvAyUZLScgrKahpaYNkwKqgs8BRgxXDmljZ2Ds619SCNzVGeAJS4AHx+DKCQ-fAAdCwgYUTk1HRMrBy0MDz8QqLi21Ky8kqq6pra22YWVrb2TtgjQ1HECMiomPArgXEwAEQUf4wZ5vRAodBYH5rXyA-7bGIcWEqYGg94Qr7QoIAoEIzZ-f4o7a7NIHP4ARhBwyi6M+UMIq2xcLxsWR8J2qX2GV85KOpVOFQu1V8CQ4VJp4Lp3wZv1huI2rIBRJgxzKZ0qlzAIpkYtBQA)

Reply:
I just changed it again sorry! i think what Max suggested initially to memoize on `getCookie` might be better because we want to memoize on fetch time. We cannot assume that a setCookie has happened in the same session as a get i think? And I changed the setCookie to override the memoized value if it exists.
But great idea with params object, could be useful if we are back to 5 params again

Reply:
https://github.com/guardian/libs/pull/162/commits/ef4c8daa1279f93ef6886209d649a12f9d307f33

Reply:
i wonder if memoizing cookie reading is really just a premature optimisation that only adds complexity? (sorry i know i brought it up!)

Reply:
maybe but it could also be really useful! hopefully it will be used by consumers. since it's only optional functionality maybe there's no harm leaving it in? it can also be refactored if needs be